lib/galaxy/model/migrate/versions/0175_history_audit.py

```diff
@@ -0,0 +1,86 @@
+"""
+Add history audit table and associated triggers
+"""
+
+import datetime
+import logging
+
+from sqlalchemy import Column, DateTime, ForeignKey, Index, Integer, MetaData, Table
+
+from galaxy.model.migrate.triggers import (
+    history_update_time_field as old_triggers,  # rollback to old ones
+    update_audit_table as new_triggers,  # install me
+)
+from galaxy.model.migrate.versions.util import (
+    create_table,
+    drop_table
+)
+
+log = logging.getLogger(__name__)
+now = datetime.datetime.utcnow
+metadata = MetaData()
+
+# NOTE: A normal incrementing PK is not typically used in an audit table,
+# just foreign keys, but the ORM will probably choke without it
+AuditTable = Table(
+    "history_audit",
+    metadata,
+    Column("id", Integer, primary_key=True),
```

Comment:
you can skip the Id column and simply mark the other 2 as a composite primary key (just add `primary_key=True` to both). Then both postgres and sqlite will automatically create an index.

Reply:
I'm glad to see the PK go away, but is creating a compound PK with one of the columns a date really better than just indexing by date desc so the MAX() select is optimized?

Reply:
I don't have any experimental data to support this, but I think we should be fine with a composite key. With a composite primary key, Postgres will automatically create a b-tree index on the 2 columns (history_id will be the leading column, both columns sorted in ascending order by default). The alternative approach would create the same b-tree index on the same 2 columns, except it would respect the `desc` directive and store its elements in descending order. However, that would only affect the case when we are interested in retrieving *multiple rows* in *that order* - then postgres (in most cases) would not need to run a sort on the result before returning it. But we only need one record (latest update_time) - and the speed of finding one element will be exactly the same, no matter in what order the elements are stored in the tree (there's no sequential scan).

Reply:
We'll need to add `ON CONFLICT DO NOTHING` (postgresql) or `ON CONFLICT IGNORE` (sqlite) as well if we use the composite PK, as I don't think we can guarantee update_time to be unique.

Reply:
@mvdbeek I agree, especially now that we may have to add more row triggers, there will be lots of duplicates. My opinion is it's not worth the brain damage but you are all free to commit to this thing.

Reply:
I just did an explain and analyze with both situations. There's no difference in the query plan for this query (which simulates our necessary join)

```
select history_id, max(update_time) as max_update_time
from history_audit 
where history_id = 2
group by history_id
```
I'll go ahead and throw in the double key and the conflict ignores.